# Цель работы: 
Научить улучшать качество модели и научиться применять алгоритмы улучшения

## Задача:
Используя один из подходящих игрушечных датасетов. Применить алгоритмы улучшения качества и оценки моделей (перекрестная проверка, решетчатый поиск, метрики модели). В моем случае я взял датасет diabetes.


Ниже приведен наш анализ и рассмотрение алгоритмов улучшения по отдельности:
(В конце нашего файла приведена новая улучшенная модель)

## Перекрестная проверка

In [68]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.datasets import load_diabetes

diabetes = load_diabetes()
logreg = LinearRegression()

scores = cross_val_score(logreg, diabetes.data, diabetes.target, cv = 5)
print("Cross-validation score: {}".format(scores))

Cross-validation score: [0.42955615 0.52259939 0.48268054 0.42649776 0.55024834]


In [69]:
print("Средняя правильность перекрестной проверки: {:.2f}".format(scores.mean()))

Средняя правильность перекрестной проверки: 0.48


In [70]:
from sklearn.model_selection import cross_validate
res = cross_validate(logreg, diabetes.data, diabetes.target, cv=5, return_train_score=True)
display(res)

{'fit_time': array([0.00200224, 0.        , 0.00099993, 0.0010004 , 0.        ]),
 'score_time': array([0.0010004 , 0.00099802, 0.        , 0.        , 0.00100183]),
 'test_score': array([0.42955615, 0.52259939, 0.48268054, 0.42649776, 0.55024834]),
 'train_score': array([0.52428257, 0.51032125, 0.52379451, 0.53137766, 0.50775068])}

In [71]:
print("Метки ирисов:\n{}".format(diabetes.target))

Метки ирисов:
[151.  75. 141. 206. 135.  97. 138.  63. 110. 310. 101.  69. 179. 185.
 118. 171. 166. 144.  97. 168.  68.  49.  68. 245. 184. 202. 137.  85.
 131. 283. 129.  59. 341.  87.  65. 102. 265. 276. 252.  90. 100.  55.
  61.  92. 259.  53. 190. 142.  75. 142. 155. 225.  59. 104. 182. 128.
  52.  37. 170. 170.  61. 144.  52. 128.  71. 163. 150.  97. 160. 178.
  48. 270. 202. 111.  85.  42. 170. 200. 252. 113. 143.  51.  52. 210.
  65. 141.  55. 134.  42. 111.  98. 164.  48.  96.  90. 162. 150. 279.
  92.  83. 128. 102. 302. 198.  95.  53. 134. 144. 232.  81. 104.  59.
 246. 297. 258. 229. 275. 281. 179. 200. 200. 173. 180.  84. 121. 161.
  99. 109. 115. 268. 274. 158. 107.  83. 103. 272.  85. 280. 336. 281.
 118. 317. 235.  60. 174. 259. 178. 128.  96. 126. 288.  88. 292.  71.
 197. 186.  25.  84.  96. 195.  53. 217. 172. 131. 214.  59.  70. 220.
 268. 152.  47.  74. 295. 101. 151. 127. 237. 225.  81. 151. 107.  64.
 138. 185. 265. 101. 137. 143. 141.  79. 292. 178.  91. 116.  8

In [72]:
from sklearn.model_selection import KFold
kfold = KFold(n_splits=5)

In [73]:
print("Значения правильности перекрестной проверки: \n{}".format(cross_val_score(logreg, diabetes.data, diabetes.target, cv=kfold)))

Значения правильности перекрестной проверки: 
[0.42955615 0.52259939 0.48268054 0.42649776 0.55024834]


In [74]:
kfold =  KFold(n_splits=3)
print("Значения правильности перекрестной проверки: \n{}".format(cross_val_score(logreg, diabetes.data, diabetes.target, cv=kfold)))

Значения правильности перекрестной проверки: 
[0.46930418 0.48725261 0.50954961]


In [75]:
kfold = KFold(n_splits=3, shuffle=True, random_state=0)
print("Значения правильности перекрестной проверки: \n{}".format(cross_val_score(logreg, diabetes.data, diabetes.target, cv=kfold)))

Значения правильности перекрестной проверки: 
[0.40409883 0.52126483 0.54380121]


## Перекрестная проверка с исключением по одному


In [76]:
from sklearn.model_selection import LeaveOneOut, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.datasets import load_diabetes

diabetes = load_diabetes()
logreg = LinearRegression()

loo = LeaveOneOut()
scores = cross_val_score(logreg, diabetes.data, diabetes.target, cv=loo)

print("Number of iterations: ", len(scores))
print("Mean accuracy: {:.2f}".format(scores.mean()))


D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.py:996: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.py:996: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.py:996: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.py:996: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.

Number of iterations:  442
Mean accuracy: nan


D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.py:996: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.py:996: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.py:996: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.py:996: UndefinedMetricWarning: R^2 score is not well-defined with less than two samples.
  warnings.warn(msg, UndefinedMetricWarning)
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\metrics\_regression.

In [77]:
import numpy as np

# Check for NaN or missing values in the dataset
nan_indices = np.isnan(diabetes.data).any(axis=1) | np.isnan(diabetes.target)
print("Number of NaN values in dataset:", np.sum(nan_indices))

Number of NaN values in dataset: 0


Перекрестная проверка со случайными перестановками при разбиении

In [78]:
from sklearn.model_selection import ShuffleSplit
shuffle_split = ShuffleSplit(test_size=.5, train_size=.5, n_splits=10)
scores = cross_val_score(logreg, diabetes.data, diabetes.target, cv=shuffle_split)
print("Значения правильности перекрестной проверки: \n{}".format(scores))

Значения правильности перекрестной проверки: 
[0.47584505 0.49577441 0.45326938 0.43618714 0.43480874 0.50569734
 0.52461098 0.52371524 0.453634   0.4326322 ]


Перекрестная проверка с использованием групп


In [79]:
from sklearn.model_selection import GroupKFold
from sklearn.datasets import load_diabetes

# Load the diabetes dataset
diabetes = load_diabetes()
X = diabetes.data
y = diabetes.target

# Creating groups based on sample indices
groups = [i // 50 for i in range(len(X))]  # Assigning a group every 50 samples

# Using GroupKFold for cross-validation
group_kfold = GroupKFold(n_splits=3)
logreg = LinearRegression()

scores = cross_val_score(logreg, X, y, groups=groups, cv=group_kfold)
print("Значения правильности перекрестной проверки: \n{}".format(scores))

Значения правильности перекрестной проверки: 
[0.41340016 0.48087214 0.56921837]


# Решетчатый поиск


In [80]:
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.datasets import load_diabetes

# Load the diabetes dataset
diabetes = load_diabetes()
X = diabetes.data
y = diabetes.target

# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
print("Размер обучающего набора: {} размер тестового набора: {}".format(X_train.shape[0], X_test.shape[0]))

best_score = 0

for gamma in [0.001, 0.01, 0.1, 1, 10, 100]:
    for C in [0.001, 0.01, 0.1, 1, 10, 100]:
        # For each combination of parameters, train an SVC
        svm = SVC(gamma=gamma, C=C)
        svm.fit(X_train, y_train)
        
        # Evaluate the SVC on the test set
        score = svm.score(X_test, y_test)
        
        # If we got a better score, store the score and parameters
        if score > best_score:
            best_score = score
            best_parameters = {'C': C, 'gamma': gamma}

print("Наилучшее значение правильности: {:.2f}".format(best_score))
print("Наилучшие значения параметров: {}".format(best_parameters))

Размер обучающего набора: 331 размер тестового набора: 111
Наилучшее значение правильности: 0.02
Наилучшие значения параметров: {'C': 0.001, 'gamma': 0.001}


In [81]:
from sklearn.svm import SVC 
# разбиваем данные на обучающий+проверочный набор и тестовый набор 
X_trainval, X_test, y_trainval, y_test = train_test_split(diabetes.data, diabetes.target, random_state=0) 
# разбиваем обучающий+проверочный набор на обучающий и проверочный наборы 
X_train, X_valid, y_train, y_valid = train_test_split(X_trainval, y_trainval, random_state=1) 
print("Размер обучающего набора: {} размер проверочного набора: {} размер тестового набора:" " {}\n".format(X_train.shape[0], X_valid.shape[0], X_test.shape[0])) 
 
best_score = 0 
for gamma in [0.001, 0.01, 0.1, 1, 10, 100]:     
    for C in [0.001, 0.01, 0.1, 1, 10, 100]: 
        # для каждой комбинации параметров обучаем SVC         
        svm = SVC(gamma=gamma, C=C)         
        svm.fit(X_train, y_train) 
        # оцениваем качество SVC на тестовом наборе         
        score = svm.score(X_valid, y_valid) 
        # если получаем наилучшее значение правильности, сохраняем значение и параметры 
        if score > best_score:             
            best_score = score 
            best_parameters = {'C': C, 'gamma': gamma} 
# заново строим модель на наборе, полученном в результате объединения обучающих
# и проверочных данных, оцениваем качество модели на тестовом наборе 
svm = SVC(**best_parameters) 
svm.fit(X_trainval, y_trainval) 
test_score = svm.score(X_test, y_test) 
print("Лучшее значение правильности на проверочном наборе: {:.2f}".format(best_score)) 
print("Наилучшие значения параметров: ", best_parameters) 
print("Правильность на тестовом наборе с наилучшими параметрами: {:.2f}".format(test_score))  

Размер обучающего набора: 248 размер проверочного набора: 83 размер тестового набора: 111
Лучшее значение правильности на проверочном наборе: 0.01
Наилучшие значения параметров:  {'C': 0.001, 'gamma': 0.001}
Правильность на тестовом наборе с наилучшими параметрами: 0.02


Решетчатый поиск с перекрестной проверкой


In [82]:
import numpy as np
for gamma in [0.001, 0.01, 0.1, 1, 10, 100]:
    for C in [0.001, 0.01, 0.1, 1, 10, 100]:
        # для каждой комбинации параметров,
        # обучаем SVC
        svm = SVC(gamma=gamma, C=C)
        # выполняем перекрестную проверку
        scores = cross_val_score(svm, X_trainval, y_trainval, cv=5)
        # вычисляем среднюю правильность перекрестной проверки
        score = np.mean(scores)
        # если получаем лучшее значение правильности, сохраняем значение и параметры
        if score > best_score:
            best_score = score
            best_parameters = {'C': C, 'gamma': gamma}
# заново строим модель на наборе, полученном в результате
# объединения обучающих и проверочных данных
svm = SVC(**best_parameters)
svm.fit(X_trainval, y_trainval)

D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated clas

SVC(C=100, gamma=0.1)

Поскольку решетчатый поиск с перекрестной проверкой является весьма распространенным методом настройки параметров, библиотека scikit-learn предлагает класс GridSearchCV, в котором решетчатый поиск реализован в виде модели.

In [83]:
param_grid = {'C': [0.001, 0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1, 10, 100]}
print("Сетка параметров:\n{}".format(param_grid))

Сетка параметров:
{'C': [0.001, 0.01, 0.1, 1, 10, 100], 'gamma': [0.001, 0.01, 0.1, 1, 10, 100]}


In [84]:
from sklearn.model_selection import GridSearchCV
from sklearn.svm import SVC
grid_search = GridSearchCV(SVC(), param_grid, cv=5)
X_train, X_test, y_train, y_test = train_test_split(diabetes.data, diabetes.target, random_state=0)
grid_search.fit(X_train, y_train)

D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


GridSearchCV(cv=5, estimator=SVC(),
             param_grid={'C': [0.001, 0.01, 0.1, 1, 10, 100],
                         'gamma': [0.001, 0.01, 0.1, 1, 10, 100]})

In [85]:
print("Правильность на тестовом наборе: {:.2f}".format(grid_search.score(X_test, y_test)))

Правильность на тестовом наборе: 0.02


In [86]:
print("Наилучшие значения параметров: {}".format(grid_search.best_params_))
print("Наилучшее значение кросс-валидац. правильности: {:.2f}".format(grid_search.best_score_))

Наилучшие значения параметров: {'C': 10, 'gamma': 1}
Наилучшее значение кросс-валидац. правильности: 0.02


In [87]:
print("Наилучшая модель: \n{}".format(grid_search.best_estimator_))

Наилучшая модель: 
SVC(C=10, gamma=1)


## Применение различных стратегий перекрестной проверки с помощью решетчатого поиска


In [88]:
scores = cross_val_score(GridSearchCV(SVC(), param_grid, cv=5), diabetes.data, diabetes.target, cv=5)
print("Значения правильности перекрестной проверки: ", scores)
print("Среднеее значение правильности перекрестной проверки: ", scores.mean())

D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated clas

Значения правильности перекрестной проверки:  [0.01123596 0.01123596 0.02272727 0.         0.04545455]
Среднеее значение правильности перекрестной проверки:  0.018130745658835545


In [89]:
def nested_cv(X, y, inner_cv, outer_cv, Classifier, parameter_grid):
    outer_scores = []
    # для каждого разбиения данных во внешней перекрестной проверке 
    # (метод split возвращает индексы)   
    for training_samples, test_samples in outer_cv.split(X, y):
        # находим наилучшие параметры с помощью внутренней перекрестной проверки  
        best_parms = {}
        best_score = -np.inf
        # итерируем по параметрам   
        for parameters in parameter_grid:
             # собираем значения правильности по всем внутренним разбиениям 
            cv_scores = []
             # итерируем по разбиениям внутренней перекрестной проверки 
            for inner_train, inner_test in inner_cv.split(
                    X[training_samples], y[training_samples]):
                # строим классификатор с данными параметрами на внутреннем обучающем наборе
                clf = Classifier(**parameters)
                clf.fit(X[inner_train], y[inner_train])
               # оцениваем качество на внутреннем тестовом наборе
                score = clf.score(X[inner_test], y[inner_test])
                cv_scores.append(score)
            # вычисляем среднее значение правильности по внутренним блокам
            mean_score = np.mean(cv_scores)
            if mean_score > best_score:
                # если лучше, чем предыдущие, запоминаем параметры 
                best_score = mean_score
                best_params = parameters
        # строим классификатор с лучшими параметрами на внешнем обучающем наборе
        clf = Classifier(**best_params)
        clf.fit(X[training_samples], y[training_samples])
        # оцениваем качество на внешнем тестовом наборе 
        outer_scores.append(clf.score(X[test_samples], y[test_samples]))
    return np.array(outer_scores)


In [90]:
from sklearn.model_selection import ParameterGrid, StratifiedKFold
scores = nested_cv(diabetes.data, diabetes.target, StratifiedKFold(5), StratifiedKFold(5), SVC, ParameterGrid (param_grid))
print("Значения правильности перекрестной проверки: {}".format(scores))

D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(
D:\Users\Bekbol2\PycharmProjects\jupyter\venv\Lib\site-packages\sklearn\model_selection\_split.py:737: UserWarning: The least populated clas

Значения правильности перекрестной проверки: [0.02247191 0.02247191 0.02272727 0.01136364 0.02272727]


## Метрики модели

In [91]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.dummy import DummyClassifier
from sklearn.metrics import accuracy_score

# Загрузка данных Iris
iris = load_iris()
X = iris.data
y = iris.target

# Разделение на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Создание и обучение DummyClassifier
dummy_clf = DummyClassifier(strategy='most_frequent')  # Стратегия выбора предсказания
dummy_clf.fit(X_train, y_train)

# Предсказание
predictions = dummy_clf.predict(X_test)

# Оценка качества модели
accuracy = accuracy_score(y_test, predictions)
print(f"Точность: {accuracy}")


Точность: 0.3


In [92]:
from sklearn.tree import DecisionTreeClassifier
tree = DecisionTreeClassifier(max_depth=2).fit(X_train, y_train)
pred_tree = tree.predict(X_test)
print("Правильность на тестовом наборе: {:.2f}".format(tree.score(X_test, y_test)))

Правильность на тестовом наборе: 0.97


In [93]:
from sklearn.linear_model import LogisticRegression
dummy = DummyClassifier().fit(X_train, y_train)
pred_dummy = dummy.predict(X_test)
print("правильность dummy: {:.2f}".format(dummy.score(X_test, y_test)))
logreg = LogisticRegression(C=0.1, max_iter=1000).fit(X_train, y_train)
pred_logreg = logreg.predict(X_test)
print("правильность logreg: {:.2f}".format(logreg.score(X_test, y_test)))

правильность dummy: 0.30
правильность logreg: 1.00


Трудно судить, какой из этих результатов является действительно полезным. Проблема здесь заключается в том, что для несбалансированных наборов данных правильность не является адекватной метрикой, позволяющей количественно оценить прогностическую способность модели

In [94]:
from sklearn.metrics import confusion_matrix 
confusion = confusion_matrix(y_test, pred_logreg) 
print("Confusion matrix: \n{}".format(confusion))

Confusion matrix: 
[[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]


In [95]:
print("\nДамми-модель: ")
print(confusion_matrix(y_test, pred_dummy))
print("\nДерево решений:")
print(confusion_matrix(y_test, pred_tree))
print("\nЛогистическая регрессия")
print(confusion_matrix(y_test, pred_logreg))


Дамми-модель: 
[[ 0 10  0]
 [ 0  9  0]
 [ 0 11  0]]

Дерево решений:
[[10  0  0]
 [ 0  8  1]
 [ 0  0 11]]

Логистическая регрессия
[[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]


## Конечная улучшенная модель используя датасет diabetes 
## на основе всех алгоритмов улучшения что мы рассмотрели выше:

In [96]:
from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split, cross_val_score, GridSearchCV
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

# Загрузка данных
diabetes = load_diabetes()
X = diabetes.data
y = diabetes.target

# Разделение на обучающий и тестовый наборы
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Инициализация модели
model = RandomForestRegressor(random_state=42)

# Перекрестная проверка (Cross-validation)
cv_scores = cross_val_score(model, X_train, y_train, cv=5)  # cv=5 означает 5-кратную перекрестную проверку
print("Средняя точность перекрестной проверки:", cv_scores.mean())

# Решетчатый поиск параметров (Grid Search)
param_grid = {
    'n_estimators': [50, 100, 150],
    'max_depth': [None, 5, 10, 15]
}

grid_search = GridSearchCV(model, param_grid, cv=5)
grid_search.fit(X_train, y_train)
print("Лучшие параметры:", grid_search.best_params_)

# Обучение модели с лучшими параметрами
best_model = grid_search.best_estimator_
best_model.fit(X_train, y_train)
predictions = best_model.predict(X_test)

# Оценка модели
mse = mean_squared_error(y_test, predictions)
mae = mean_absolute_error(y_test, predictions)
r2 = r2_score(y_test, predictions)

print(f"Среднеквадратичная ошибка: {mse}")
print(f"Средняя абсолютная ошибка: {mae}")
print(f"Коэффициент детерминации (R^2): {r2}")


Средняя точность перекрестной проверки: 0.39093177542361224
Лучшие параметры: {'max_depth': 5, 'n_estimators': 150}
Среднеквадратичная ошибка: 2886.9282789123713
Средняя абсолютная ошибка: 43.492022897852586
Коэффициент детерминации (R^2): 0.45510653225048425


Этот код использует RandomForestRegressor в качестве модели, выполняет перекрестную проверку для оценки точности модели, а затем выполняет решетчатый поиск для подбора наилучших гиперпараметров модели. Затем оцениваются метрики качества, такие как среднеквадратичная ошибка (MSE), средняя абсолютная ошибка (MAE) и коэффициент детерминации (R^2) на тестовом наборе данных.

# Вывод:

Мы рассмотрели алгоритмы улучшения качества и оценки моделей (перекрестная проверка, решетчатый поиск, метрики модели) на основе датасета diabetes. Поняли что не все алгоритмы подходят для всех датасетов, но можно отлично подобрать под свой. И тем самым улучшить модель.

В случае с нашим датасетом diabetes, в 5 лабараторной работе мы видели правильность(точность) от 26% до 38%, а здесб мы смоли добиться 45%.